### PR TITLE
Use StackedSparseArray for MetadataMatch `sparse equal_match`

### LINKED ISSUE #544: MetadataMatch with "sparse" flag still allocates whole matrix
**Describe the bug**
If the user passes the `sparse` flag, it does maybe not make sense to allocate the whole matrix but to compute each entry separate and then add them to the sparse array.

### PR DESCRIPTION
**Pull request by zargham-ahmad**
This PR includes the following:

- slightly refactored and updated MetadataMatch matrix function to also use StackedSparseArray for `equal_match` matching_type when array_type is `sparse`

closes #544

### CODE REVIEW

**Review by niekdejonge — APPROVED** [review]
Looks good to me @zargham-ahmad thanks!

**Comment by niekdejonge** [comment]
Not completely related to this PR, but I did not look at the MetadataMatch class before, the difference between "equal_match" and "difference" was not fully clear to me at first glance. Can you maybe add a bit of extra docstring here @zargham-ahmad? And maybe we should also give a warning if a tolerance is given, but matching_type is set to "equal_match".

File: matchms/similarity/MetadataMatch.py [hunk 1]
```diff
@@ -130,6 +130,9 @@ def matrix(self, references: List[SpectrumType], queries: List[SpectrumType],
             comparison). By using the fact that score[i,j] = score[j,i] the calculation will be about
             2x faster.
         """
+        if array_type not in ["numpy", "sparse"]:
+            raise ValueError("array_type must be 'numpy' or 'sparse'.")
+
```
Comment (hechth) [inline matchms/similarity/MetadataMatch.py:136]: Good addition - this should be covered by a unit test though to make sure this specification is tracked.

File: matchms/similarity/MetadataMatch.py [hunk 2]
```diff
@@ -150,25 +153,28 @@ def collect_entries(spectrums):
...
         if array_type == "sparse":
             scores_array = StackedSparseArray(len(entries_ref), len(entries_query))
             scores_array.add_sparse_data(rows, cols, scores.astype(self.score_datatype), "")
-            return scores_array
-        raise ValueError("")
+        else:
+            scores_array = np.zeros((len(entries_ref), len(entries_query)), dtype=self.score_datatype)
+            scores_array[rows, cols] = scores.astype(self.score_datatype)
+
+        return scores_array
```
Comment (hechth) [inline matchms/similarity/MetadataMatch.py:185]: Is there any change made to the return type of the function based on the changes in this PR?

Comment (zargham-ahmad) [reply matchms/similarity/MetadataMatch.py:185]: No, as for the difference method the current return type was used, and the matrix function in the BaseSimilarity also has the same return type

File: matchms/similarity/MetadataMatch.py (hunk 1 above)
Comment (zargham-ahmad) [reply matchms/similarity/MetadataMatch.py:136]: test added in [24a8ecb](https://github.com/matchms/matchms/pull/642/commits/24a8ecb2aa784fb4e8ff9ab69464138c48b70ad9)

**Comment by zargham-ahmad** [comment]
> Not completely related to this PR, but I did not look at the MetadataMatch class before, the difference between "equal_match" and "difference" was not fully clear to me at first glance. Can you maybe add a bit of extra docstring here @zargham-ahmad? And maybe we should also give a warning if a tolerance is given, but matching_type is set to "equal_match".

done in [24a8ecb](https://github.com/matchms/matchms/pull/642/commits/24a8ecb2aa784fb4e8ff9ab69464138c48b70ad9)

**Comment by niekdejonge** [comment]
@zargham-ahmad Thanks for the additions. Is it ready to be merged? If so feel free to merge, or if you don't have the rights, let me know :)

**Comment by zargham-ahmad** [comment]
> @zargham-ahmad Thanks for the additions. Is it ready to be merged? If so feel free to merge, or if you don't have the rights, let me know :)

Yes, it is ready. Please merge it.

**Pull request merged**